packages/nimble-components/src/rich-text/specs/mention-hld.md

```diff
@@ -0,0 +1,617 @@
+
+Validation will be integrated into the internal workings through a `validConfiguration` flag. There is a base class called `RichTextMentionValidator`
+responsible for handling validation, with methods for setting and retrieving the validation configuration. This approach ensures that the state of
+valid and invalid values passed within the mapping element is effectively managed. The class obtains the name of the validity flag to communicate this
+information via a public API called `validity`.
+
+By deriving from the base, the mention options can validate the following conditions for the `key` and `display-name` value in mapping element:
+
+1. `validateMappingTypes(mappings)`
+2. `validateUniqueKeys(keys, keyType)`
+3. `validateNoMissingKeys(mappings)`
+4. `validateNoMissingText(mappings)`
+
+_Note_: These are subject to change based on the property changes in the `nimble-mapping-mention-user` element.
+
+If any of the mention option is invalid as per the above validation, that particular option will render as an empty
+option in the list. This indicates the client that some of the option is wrongly configured and identify the
+validation details using the public API `validity`.
+
+The public API to determine the validity of the mention options are `checkValidity()` and `validity`.
+See the [API section](./README.md#api) for more details.
+
+#### 6. _nimble-rich-text-user-mention-view_:
+
+The foundation for configuring the rendering of mention nodes in the UI is provided by a base class known as `nimble-rich-text-mention-view`. This base class is an extension of
+`FoundationElement`.
+
+The `nimble-rich-text-user-mention-view` is a view class that is derived from the above base class, and its template contains a `span` element with the required CSS styling.
+They are equipped with attributes such as `mention-id`, `mention-label`, and `mention-type` to render the node as a `Mention` node in Tiptap editor.
+It uses the default slot to render the same text content from the editor to the view element's shadow root.
+
+The Mention node from Tiptap is extended, and the [`renderHTML`](https://tiptap.dev/guide/custom-extensions/#render-html) method is overridden to render
+the element as a `nimble-rich-text-user-mention-view` instead of the default `span` element as shown in the
+[Tiptap code](https://github.com/ueberdosis/tiptap/blob/42039c05f0894a2730a7b8f1b943ddb22d52a824/packages/extension-mention/src/mention.ts#L112).
+The attributes will be retained and included in the rendered mention nodes. The [`parseHTML`](https://tiptap.dev/guide/custom-extensions/#parse-html)
+will also hold the `nimble-rich-text-user-mention-view` tag and `span` with type attribute mention to load the specified tags as mention node
+in the editor.
+
+#### 7. _nimble-rich-text-mention-list-box_:
+
+This component extends FAST's `list-box` to inherit the essential mouse and keyboard interactions required within the dropdown list. Within the default slot, it gets
```

Comment:
Do we need to extend the list box or could this be a custom element that contains a list box in its template? The latter is preferable since it only relies on FAST's public API so it's less likely to break if they change the internals of the custom element. It also gives us more flexibility if we want to add content around the listbox.

Reply:
@jattasNI We have to extend the list box to get the necessary state like the selected index value to navigate between options. But if we are required to use the list box within a template, it can only be done in the below ways,
1. The children of `nimble-rich-text-mention-list-box` should be `nimble-list-option` to get the user information. In the template, by using the children's reference, we could get the information and again render the `nimble-list-option` (not via a slot) with all the click and key-down handlers for the popup along with the filter logic encapsulated within this component. This is because the `nimble-rich-text-mention-list-box` will be removing the list options based on filtering inside its own shadow root.
2. Have just a `nimble-list-box` component (as they are not available in the nimble) and handle all the filtering logic, click handlers, and key down handlers within the editor and dynamically populate the `nimble-list-option` based on filtering. 

Option 1 is what we have in the [current prototype branch](https://github.com/ni/nimble/pull/1593/files#diff-cd27bfa506b071144ae2da11fadd04eb71dc27396ae478e100318bf4ee6d3671) which renders the list box and list options within the custom mention list box element. However, it seems like we are rendering the list options twice, one is to pass the information as a child from the editor to the custom list box element and the other one is to render it in the popup.

Passing within a slot doesn't work as we will hide the slotted elements dynamically based on filtering which will not invoke the `slottedChanged` in the list box every time and so the arrow keys will not work as expected because the element is still there in the DOM and hidden case were not handled in the FAST's list box.

If we still have to use the list box in the template, we could prefer using Option 2 and avoid having a custom element for the dropdown and handle all the filtering within the editor itself. (If we prefer this option, I have to update the doc)

The initial idea for having a custom component for the dropdown is to extend the list box and easily override the list box methods based on our needs. But using a list box within a template gives more advantages as you mentioned. So we ended up with the above options. So please let us know your thoughts/suggestions.

Reply:
@m-akinc graciously agreed to look more closely at the options and help make a recommendation

Reply:
I like the idea of having a dedicated `nimble-rich-text-mention-list-box` element to encapsulate some of the related behaviors (e.g. event handling), so your option 1 seems like a good approach to me. The only drawback seems to be that, "it seems like we are rendering the list options twice." Yes, the full set of list options is rendered as children of the `nimble-rich-text-mention-list-box`, and that in turn duplicates a filtered subset of those options in its shadow DOM, as children of the `nimble-list-box`. I don't see the harm in that, as long as the number of displayed (filtered) options are limited. I.e. if there are 1000 list options, and we re-render all 1000 in the popup, that could be a problem. But if we limit to, say, 20, then I don't think it is a concern.

Another thought is that you might be able to use a slot instead. You say that FAST's arrow key handlers do not take into account if an option is hidden. However, [they _do_ take into account if the option is disabled](https://github.com/microsoft/fast/blob/21f81852e54a7ee2fb1a4911ff524c77a1be961f/packages/web-components/fast-foundation/src/listbox/listbox.ts#L442). It should be possible to both hide _and_ disable filtered options to get the keyboard navigation to work properly. It may be worth trying this out, but otherwise, option 1 is already a decent choice.

Reply:
Thanks, @m-akinc  for your input. Doing both _hide_ and _disable_ filtered options makes the keyboard navigations work as expected. So we can go with the slot approach which makes us have a separate component for the popup as well and the list options will not be rendered twice in the DOM.